Breaking: Calculate leading/trailing comments in core

kaicataldo opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[X] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (http://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
This PR turns off comment attachment in Espree and moves comment getting logic into `sourceCode.getComments()`. This is a breaking change.

**Is there anything you'd like reviewers to focus on?**
Would love suggestions for how else we might be able to handle shebang comments.

As discussed on the [corresponding issue](https://github.com/eslint/eslint/issues/6724), we should discuss if we want to continue thinking about comments the same way now that we're not attaching at the parser level. This PR mimics the current attachment strategy in Espree as close as it can, though it's not possible (nor do I think we want it to be) exactly the same as it is in Espree, because there are some unpredictable edge cases and bugs in that.

The version in this PR should essentially work the same for our users and ecosystem (unless they rely on some of the really weird edge cases mentioned before).

nzakas commented on lib/eslint.js:
```diff
@@ -908,6 +908,8 @@ module.exports = (function() {
                 }
             }
 
+            ast.hasShebang = !!shebang;
```
Hmm, this doesn't look like a good idea.

kaicataldo replied:
Yeah, this was what I was hoping to get some feedback on. The current behavior is to remove the shebang from the `comments` array and from the `leadingComments` of the first node in the Program body after parsing has completed. It also modifies the comment to look like a standard JS line comment. Now that we're calculating this on the fly, I need to figure out how `getComments()` can know which token represents a shebang comment (if one exists) so that it doesn't include it. Is there any possibility of adding a property to the token?

kaicataldo replied:
Here's the line where the `SourceCode` instance is created: https://github.com/eslint/eslint/blob/master/lib/eslint.js#L817

And here's where the shebang comment is removed: 
https://github.com/eslint/eslint/blob/master/lib/eslint.js#L903

kaicataldo replied:
That's right - sorry I'm not explaining this better. The `SourceCode` instance above takes the given `ast` from parser and generates its own internal `tokenAndCommentStore` from it. Since this occurs before the shebang comment is removed, the `SourceCode` instance's `tokenAndCommentStore` still contains the shebang comment.

I think we should be able to do the shebang comment removal before creating the `SourceCode` instance - will have to figure out how to do it with the few forks of logic that happen there.

kaicataldo replied:
After a bit of digging I'm realizing that including the shebang in the tokens and comments store might be intentional and actually a good thing. Despite not being attached to any nodes or included in the `comments` array of the AST, transforming it into a standard JS line comment and including it in the store allows us to use `sourceCode.getTokenOrCommentBefore()` in rules to write rules around the shebang.

Had some thoughts and will write them in a comment below.

nzakas commented on lib/util/source-code.js:
```diff
@@ -133,6 +143,9 @@ function SourceCode(text, ast) {
     this.getTokenOrCommentAfter = tokensAndCommentsStore.getTokenAfter;
 
+    this._getTokens = tokensAndCommentsStore.getTokens;
+    this._getCommentsStore = new WeakMap();
```
Maybe just `_commentStore`?

kaicataldo commented:
Sure, sorry that wasn't clear. 

### Differences between `getComments()` and Espree's comment attachment
* `getComments()` iterates over token lists and will always stop when it encounters a non-comment token (starting from the first token and checking before and the last token and checking after). This is different from the current behavior, as Espree collects comments as it parses and then attaches when it finishes a node. This means that comments can be attached across tokens and leads to some unexpected behavior (see examples below).

* It doesn't attach nodes that exist outside the range of the node's parent, as it should then be attached to that parent node (I didn't notice any differences between Espree and this PR for this behavior).

### Examples:
```js
foo /*comment*/ || /*comment*/ bar
var foo /*comment*/ = /*comment*/ bar;
```
**Espree:** In both examples above, the Identifier`foo` has 1 trailing comment while `bar` has 2 leading comments.

**sourceCode.getComments():** In both examples above, the Identifier`foo` has 1 trailing comment and the Identifier`bar` has 1 leading comments.

```js
function foo(/*asdf*/) {}
function foo(/*asdf*/bar) {}
```
**Espree:** In the first example above, the BlockStatement has a leading comment. In the second, the Identifier `bar` has a leading comment and the BlockStatement doesn't have any.

**sourceCode.getComments():** The first example does not return any comments. I wasn't sure what the desired behavior would be here - should it attach the comment as a trailing comment when a function node's `params` is empty? In the second, the Identifier `bar` has a leading comment. The BlockStatement does have leading comments in either case.

### Questions/Concerns
* How should we handle shebang comments (since they shouldn't be included in the results for `getComments()`?
* How should an empty function node params comments be treated? This is where the model of leading/trailing comments starts making less sense. I don't think it makes sense for them to be attached to the function body as a leading comment.
* How does everyone feel about the slightly changed (and more predictable) behavior described above?

nzakas commented:
Thanks, that's super helpful. I think the new behavior you've described makes a lot of sense, and have no objections to either dropping the comment without a node or the slightly changed attachment behavior of leading comments.

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,7 +961,110 @@ describe("SourceCode", () => {
+            const code = [
+                "//#!/usr/bin/env node",
+                "var a;",
+                "// foo",
```
Does this comment count as trailing for `var a;` *and* leading for `var b;`?

I don't think this is a problem since people who want to iterate over all comments can just iterate over the comment store without worrying about attachment, but I also want to make sure I understand what's going on here.

Maybe some comments around the asserts would help? (E.g., `assertCommentCount(1, 1)(node); // commented shebang, foo`)

platinumazure replied:
I'm okay with the behavior-- I just wanted to confirm my understanding.

I'd love to see some comments in the tests themselves, so it's clear to people unfamiliar with comment attachment which comments are leading and trailing to what nodes. 90% of the time it's clear, but for the other 10%, it'd be nice to see documentation via comments.

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,7 +961,110 @@ describe("SourceCode", () => {
+            eslint.on("Identifier", assertCommentCount(0, 0));
+
+            eslint.verify(code, config, "", true);
+        });
```
Could you please add a test for a multiple-declarator VariableDeclaration, so we understand how the comment attachment is supposed to work there?

Example test case:

```js
// Leading comment for VariableDeclaration?
var a,  // Trailing comment for VariableDeclarator? And/or leading for the next?
    b,  // Trailing comment for second VariableDeclarator?
    c;  // Trailing comment for VariableDeclaration?
// Trailing comment for VariableDeclaration?
```

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -940,7 +1076,352 @@ describe("SourceCode", () => {
+                "switch (foo)",
+                "    //comment",
+                "    /*another comment*/",
+                "}"
```
Is this example even syntactically valid? I see a closing brace but not an opening brace.

kaicataldo replied:
Good catch :+1:

kaicataldo commented:
Working on this has led me to have some questions around how we want to handle shebangs. Essentially, it seems like we actually probably want to keep the current behavior of not including shebangs in the AST's `comments` array and should not be included as a leading comment when we use `sourceCode.getComments()`. Please see [this comment thread](https://github.com/eslint/eslint/pull/7516#discussion_r86607115) for more context.

The problem as it currently stands is that `sourceCode.getComments()` doesn't have a way of knowing if a LineComment token was a shebang or not (since it gets transformed into a standard JS LineComment prior to parsing). I think we do want to keep the behavior of transforming the shebang and keeping it in the tokens and comment store, as this allows rules to use `sourceCode.getTokenOrCommentBefore()` to get the token that represents the shebang.

It seems like we have a few ways forward, and I wanted to see what you all thought:
* Right after parsing in lib/eslint.js, add a property (`shebang: true`?) to the LineComment token that represents the shebang and then not filter it out in `sourceCode.getComments()`'s return value.
* When `sourceCode` is instatiated, remove the shebang and store a property/add a helper method in `sourceCode` that allows access to the shebang token.
* Simply remove the shebang token and make rules figure this out for themselves (maybe by checking the first line of the source code). This is my least favorite option, because it feels like something we should be able to provide.

Thoughts? Suggestions? Things I missed?

platinumazure commented:
I'd vote for option 2:

> * When sourceCode is instatiated, remove the shebang and store a property/add a helper method in sourceCode that allows access to the shebang token.

This is similar to how we handle the byte-order mark (BOM).

I would be okay with option 1 (add `shebang` property to the comment token itself). I would be opposed to option 3 (remove the comment entirely from SourceCode's store).

mikesherov commented:
This sounds correct to me too:

> When sourceCode is instatiated, remove the shebang and store a property/add a helper method in sourceCode that allows access to the shebang token.

kaicataldo commented:
Updated - thoughts on this approach? I thought about it some more and am actually uncomfortable with removing it from the token list, as this has potential to break rules that currently rely on shebangs being represented as line comment tokens. 

Treating the shebang like a line comment works for most cases - it's really only for cases that specifically need to know whether the line comment token it is checking represents a shebang that we don't cover at the moment. Adding the property to the token will continue to allow iterating over tokens (which seems the most likely way that rules will be checking this).

Another thought I had was changing the `type` of the comment token to `Shebang`. This could potentially break some rules that rely on checking for token `type === "Line", but would allow the token to not be removed.

Thanks for all the input!

nzakas commented:
I think that's fine. We should probably document the behavior somewhere, though, as this is subtle enough that I'm not sure anyone will be able to figure it out just be looking at the code.

kaicataldo commented:
I believe I have addressed all the above comments. As I was writing out the documentation, I realized that it is a bit odd that we don't attach the shebang comment when calling `sourceCode.getComments()`, but that it's still included in the results of `sourceCode.getAllComments()` and `sourceCode.getTokenOrCommentBefore()`.

The current behavior (before this PR) is that `sourceCode.getTokenOrCommentBefore()` is the only way to access shebangs, because they are represented by "Line" comment tokens in `sourceCode.tokensAndComments`. `sourceCode.getComments()` and `sourceCode.getAllComments()` do not return the shebang, as the comment token is removed the from leading comments of the first child node in the body as well as the top level comments array.

Any thoughts on how we can make this clearer and improve the consistency? In both the current and proposed behavior in this PR, the inconsistency of whether a shebang constitutes a comment or not is confusing. 

## Some ideas: 
### Would it be better for shebangs to be accessed via a helper method through `sourceCode`?  

I'm not convinced that's helpful, though, since when you're checking for spacing around nodes you generally iterate over tokens. This could affect the `lines-around-comment` and `lines-around-directive` rules, which rely on `sourceCode.getTokenOrCommentBefore()` to ensure proper spacing even with shebangs, and could become an issue for a rule such as the one being discussed in [this proposal](https://github.com/eslint/eslint/issues/7356).

### Should we include it in all the methods that return comments?
Feels like it would be easier for the rule writer to to be able to access it as a comment token, but then I wonder if we include it in all the methods that return comments, for the sake of consistency and ease of use. This makes it very easy for whitespace-related rules, as they could then always use `sourceCode.getComments(node)`. Now that it has a unique type, it's easier to filter out if need be and may not be an issue (let me know if I'm missing something).

kaicataldo commented:
@platinumazure That's right! Added the corresponding test. Thanks for reviewing this :)

platinumazure commented:
I'm honestly not sure what to do about the shebang. I like the idea of just treating it like any other comment (as you noted, the type is now unique so it could be filtered out if needed). But I don't feel qualified to answer, to be completely honest.

kaicataldo commented:
@platinumazure Thanks for weighing in, regardless! I think treating it like any other comment would make things easier for rule writers - all most rules care about is the location of the shebang/comment so that they can enforce whether there is whitespace around them.

There might be some rules that will need to be modified (`spaced-comment` and `capitalized-comments` come to mind), but now that the shebang comment has the type `Shebang`, they would be easy to filter out. Though this would require a little more work on the part of the rule writer, this case seems to be a lot less common than rules that would only care about the shebang's location rather than its value.

not-an-aardvark commented:
> but now that the shebang comment has the type Shebang, the shebang token would be easy to filter out.

This is only loosely related, but would it be possible to have a consistent flag or function to detect whether a given thing is a comment? When getting tokens/comments from `sourceCode.tokensAndComments`, it's not immediately obvious how to distinguish a token from a comment. I had previously been using `if (thing.type === "Line" || thing.type === "Block")` to detect comments, but it looks like that won't work anymore if we add a new `Shebang` type.

kaicataldo commented:
@not-an-aardvark Yeah, definitely possible to make a helper method. The check up to this point is exactly as you described - to keep the same behavior with this PR it would just be `token.type === "Line" || token.type === "Block" || token.type === "Shebang"`.

nzakas commented:
:+1: for treating shebang comments like any other comments. It seems like the most straightforward and expected approach. The one thing we need to be careful of is using `sourceCode.getText()` on that comment, as I think it will return `//` instead of `#` at the beginning right now.

kaicataldo commented:
@nzakas Looks like we're okay, as `sourceCode` is instantiated with the original text (we transform the first two characters of a shebang for the parsing only), and that's what's used in `sourceCode.getText()`. Thanks for pointing that out - I've included a test to assert this behavior.

kaicataldo commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -261,6 +261,8 @@ Once you have an instance of `SourceCode`, you can use the methods on it to work
 * `getTokensAfter(nodeOrToken, count)` - returns `count` tokens after the given node or token.
 * `getTokensBefore(nodeOrToken, count)` - returns `count` tokens before the given node or token.
 * `getTokensBetween(node1, node2)` - returns the tokens between two nodes.
+* `getTokenOrCommentAfter(nodeOrToken)` - returns the first token or comment after the given node or token.
```
Question: These are undocumented but are used in a few places in our own codebase. Should I include them in the docs? Or do we really not want these to be public?

kaicataldo replied:
I've moved the discussion of how we want to handle these methods to https://github.com/eslint/eslint/issues/7810.

If we can land that first, I can update this PR accordingly and remove these two methods from the docs in this PR.

nzakas replied:
I think we just forgot to update the docs, we didn't intend to keep them private.

kaicataldo replied:
I'm actually going to remove these changes since the [TokenStore rewrite](https://github.com/eslint/eslint/pull/7936) updates this documentation.

kaicataldo commented:
I need to give this one more look over with a fresh mind, but I believe I've addressed all the comments above. Shebangs are now treated as any other comments, including firing a `ShebangComment` event as traversal occurs.

@not-an-aardvark How do you feel about the addition of `isComment` to the `ast-utils`?

@eslint/eslint-team Wouldn't mind some more eyes on this to make sure I'm not missing anything/breaking something I shouldn't be :)

kaicataldo commented on lib/util/source-code.js:
```diff
@@ -191,25 +202,61 @@ SourceCode.prototype = {
+        ) {
+            const tokens = this._getTokens(node);
+
+            comments.trailing = tokens.filter(astUtils.isComment);
```
Adding comments as leading comments to an otherwise empty "Program" node but adding them as trailing comments to other empty nodes is a holdover from Espree. Is this something we'd like to keep? I didn't change it in my first pass at this to keep the behavior as similar as possible, but this seems highly like to cause confusion in the future...

kaicataldo replied:
Adding comments as leading comments to an otherwise empty "Program" node but adding them as trailing comments to other empty nodes is a holdover from Espree. Is this something we'd like to keep? I didn't change it in my first pass at this to keep the behavior as similar as possible, but this seems highly like to cause confusion in the future...

platinumazure replied:
@kaicataldo What are you proposing instead? I can't quite tell from this comment.

nzakas replied:
I think the current logic is correct. You can't logically have something trailing an empty program, so leading is the least confusing option. For other empty nodes, I'm thinking that's mostly BlockStatement? Didn't we decide to just drop those from the whole leading/trailing paradigm?

kaicataldo replied:
Yes, it's mostly empty blocks/arrays. We only talked about dropping comments in the case of a function argument's parens only containing a comment.

nzakas replied:
Ah right. I'd say we should treat all of those situations the same, so anytime there's a only a comment inside of some containing structure, we shouldn't attach it as leading or trailing.

kaicataldo replied:
Thanks - makes sense to me!

kaicataldo replied:
Worked on this today and I'm actually not sure we want to do this. There are some rules - such as `lines-around-comments` - that rely on visiting comment nodes, and ignoring these cases means that the comments in the following cases are not evaluated (and the tests fail): https://github.com/eslint/eslint/blob/master/tests/lib/rules/lines-around-comment.js#L921-L938.

There are a few other rules with failing tests (no-empty and no-empty function), however these rules can be fixed by using token and comment iterators. I'm not sure what we do for the above case - the benefit of the way Espree handled this is that these comments didn't get lost, even if it led to some unpredictable behavior.

I'm a little bit at a loss right now as to what the best way forward is. I think the behavior we discussed above for `sourceCode.getComments()` makes a lot of sense - the issue is that we also use that method to find comment tokens to visit. I wonder if this could be fixed by breaking the tie between `CommentEventGenerator` and `sourceCode.getComments()`?

I'm not sure of the the performance implications, but one thought I had was that we could iterate over all the tokens and comments of a given node in the `CommentEventGenerator` method, collect the comment tokens, and emit events for those comments. We'd still have to figure out a way to avoid emitting duplicate events as the traversal occurs, though.

kaicataldo replied:
Another idea I just had is that we could use `sourceCode.getAllComments()` to get all the comments, create some kind of store in `CommentEventGenerator`, and then check the store and emit events for comments that fall within the currently visited node's range? Sorry, I know this isn't very fleshed out. I can spend some time playing around with these and see if I get anywhere.

not-an-aardvark commented:
@not-an-aardvark How do you feel about the addition of isComment to the ast-utils?

Seems fine to me. I am slightly concerned that this function wouldn't be accessible to external rules -- ideally, I think it would be nice if external rules could have an authoritative "is this thing a comment" function that wouldn't break if we added more types of comments. But I think this is probably fine for now.

mikesherov commented:
@kaicataldo, I was under the impression that this change wouldn't change anything about the emission of events so that using the LineComment or BlockComment visitors wouldn't change.

Is there some sort of guaranteed ordering of comment events (other than after Program:start and before Program:end) that precludes from just linearly iterating over all comments and emitting events for them right after Program:start? Or are there other technical reasons that isn't achievable that I'm unaware of?

kaicataldo commented:
The [current implementation](https://github.com/eslint/eslint/blob/master/lib/util/comment-event-generator.js#L94) calls `sourceCode.getComments()` on each node while it traverses. I like the idea of iterating over the top level comments array, as this will ensure every comment is accounted for. From my understanding, this seems technically feasible, but I'm not sure if the order in which these events gets fired matters for this case or not. I'm inclined to think that it doesn't matter too much for comments.

nzakas commented:
The intent with `LineComment` and `BlockComment` was to have these emitted inline with the traversal of the AST so you could, for example, create a rule that uses something like `/*eslint-disable*/` and `/*eslint-enable*/` over a particular area of the code. IMHO, that capability is important to maintain.

kaicataldo commented:
Ah, got it. Will keep thinking about this then.

kaicataldo commented:
@nzakas I think creating a rule that uses enable/disable comments could still be possible without caring about comment traversal order. One could use `sourceCode.getAllComments()` to get all the comments, iterate over the array and find the ranges between the comments. It could then check that whatever node/token it's checking falls within that range or not and ignore it. 

This could allow us to do what @mikesherov suggested, which ensures that all comments fire an event, which I think is really important. I'm not sure how else we can ensure that we traverse through all comments, since they can anywhere. Espree got around this by always attaching comments, even if they were attached to the wrong node.

Thoughts?

nzakas commented:
@kaicatalod I don't see how we can change `LineComment` and `BlockComment` to fire outside of the order they currently do. That would be a significant breaking change and I'm not sure there's a good enough reason to do that. There's nothing stopping people from manually grabbing `Program.comments` if they want to do a straight iteration, so I'd rather keep the events firing in the same order they do now.

mikesherov commented:
Yeah, if order there is important, we definitely need a different way. Wasn't considering the case @nzakas had suggested.

platinumazure commented:
Is the "Note:" in [this section of our rules guide](http://eslint.org/docs/developer-guide/working-with-rules#accessing-comments) a sufficient CYA to just deprecate and remove support for emitting those events, maybe doing a graceful deprecation over a couple of major releases?

kaicataldo commented:
@nzakas Alright. The problem we need to solve then is how we ensure that all comments are visited (rules such as https://github.com/eslint/eslint/blob/master/lib/rules/spaced-comment.js rely on this).    We can approximate it in the `getComments()` algorithm for more common comment locations, but since they can occur anywhere (including in between tokens of a given node) it's hard to get around the lossy nature of doing this post-parse.

One potential solution could be to keep a tally of which comments have been visited using `getComments()` and then emit events for any remaining missed comments at the very end, but that kind of defeats the purpose (since they're still not being emitted in the proper order).

nzakas commented:
Another possible solution is just to remove LineComment and BlockComment as events altogether and force people to figure out their own way of working with comments. My biggest concern is about changing the behavior of something that already exists, but removing it eliminates confusion and still lets people do what they want.

kaicataldo commented:
I can get behind that (though I still do like @mikesherov's suggestion, as I think it provides a more intuitive API when dealing with comment-specific rules). 

I don't want to derail this discussion, but if we're considering removing emitting comment events altogether and are also considering adding APIs to access comments during token traversal (https://github.com/eslint/eslint/issues/7810), I question whether we need `sourceCode.getComments()` at all. Seems redundant (and confusing, because you have to understand the algorithm) - would it not be more intuitive to simply call `sourceCode.getTokenOrComment(Before|After)()`?

not-an-aardvark commented:
If it's useful as a datapoint, I never use `sourceCode.getComments(node)` because I've always found the `leading`/`trailing` paradigm unintuitive. I always use either `sourceCode.getTokenOrComment(Before|After)` or `sourceCode.ast.comments`.

nzakas commented:
> I question whether we need sourceCode.getComments() at all. 

Maybe not, but I don't think we can remove it at this point, that just changes too much at once. We can deprecate it and maybe add getCommentBefore\After to provide the same functionality (getTokenOrComment isn't quite correct as you'll always need to check if the returned object is a comment or not). Then we can remove it at a later time.

kaicataldo commented on lib/rules/no-irregular-whitespace.js:
```diff
@@ -65,6 +62,7 @@ module.exports = {
         const skipTemplates = !!options.skipTemplates;
 
         const sourceCode = context.getSourceCode();
+        const commentNodes = sourceCode.getAllComments();
```
This rule was using comment order traversal to collect comment tokens, but didn't need to. Its goal is to check all comments, so instead of collecting them while traversing we can simply use `sourceCode.getAllComments()`.

kaicataldo commented on lib/rules/lines-around-comment.js:
```diff
@@ -178,12 +178,7 @@ module.exports = {
-            if (ancestors.length) {
-                parent = ancestors.pop();
-            }
+            const parent = sourceCode.getNodeByRangeIndex(node.range[0]);
```
`context.getAncestors()` no longer works for this case because the comments are no longer included in traversal.

kaicataldo commented:
I've updated the PR to reflect the discussion above (no longer emitting events for comments). I had to update all the rules that rely on this behavior, but was able to get them all working. Thoughts?

kaicataldo commented:
### As requested, here's a summary of the scope of changes contained in this PR:

#### Calculating leading/trailing comments
The first thing this PR does is turn off comment attachment in Espree, instead implementing its own token traversal algorithm to calculate leading and trailing comments around a given node. It was not possible to emulate Espree's (buggy) behavior.

#### Treating Shebangs as any other comment token
Shebangs are now comment tokens with type `Shebang`. Shebangs are currently only accessible through what feel like arbitrary/unrelated methods at the moment, and treating them like any other comment will hopefully make it both clearer how rule authors can access/account for them as well as make it easier to work with them.

#### Removing the firing of comment events during traversal
The current comment event firing logic uses `sourceCode.getComments()` to attach comments on nodes and then emit events for them. Due to the now lossy nature of comment attachment (Espree did not lose comments because it would collect all comments as it parsed and would do its best to attach them to the corresponding node. While this had some weird edge cases and was pretty buggy, it did ensure that all comments were accounted for), we are no longer guaranteed to emit events for all comments. This is detrimental for rules like [lines-around-comment](http://eslint.org/docs/rules/lines-around-comment), as they check all comments.

As you can see from the discussion above, we weighed a few alternative ways of emitting events for all comments. We ended up deciding that it was better to remove this behavior rather than keep it and change the order in which events were fired. 

From what I've seen, most use cases for accessing comment tokens falls into one of two categories:
 1. Checking for comments around specific nodes 
 1. Checking all comments in the program for a property/style

I've added to [the docs](https://github.com/eslint/eslint/pull/7516/files#diff-31ef8f31971065bcc300804ca4d9661fR346) explaining how to do the equivalent using `sourceCode` methods.

This PR in combination with the [TokenStore rewrite](https://github.com/eslint/eslint/pull/7936) will give rule authors a lot more control over dealing with comment tokens. I also like that that we think of comments in the context of token traversal rather than AST traversal!

@eslint/eslint-team Please let me know if anything isn't clear! I'm feeling pretty good about where this stands right now, but always happy to discuss/improve it!

nzakas commented:
That's a great summary, thanks! I'm pretty happy with where this PR ended up and really appreciate all the work you put into it. Unless others have objections, I think we should start planning the next major release to get this in (so the code doesn't get stale sitting here).

kaicataldo commented:
Great! The last thing I'd like to do with this PR is rebase it off of https://github.com/eslint/eslint/pull/7936 when it lands.

kaicataldo commented:
@mysticatea Sorry to ping you about this, but do you have any idea what I might be doing wrong? 

The problem seems to occur when a switch statement is directly followed by another comment. Example:
```js
function a() {
    switch (a) {
        case b:
            return c;
        // no default
    }
    // some other comment
}
```

[As `sourceCode.getComment()` is iterating over the tokens](https://github.com/eslint/eslint/pull/7516/files#diff-4625bd821401815c773e90e5cecdcd73R269), it seems to skip the closing brace. So the returned tokens are:
```js
[{ 
    type: 'Line',
    value: ' no default'
}, { 
    type: 'Line',
    value: ' some other comment'
}]
```

I looked at this both last night and tonight and can't seem to figure out what I'm doing wrong. Any insight would be appreciated!

mysticatea commented:
@kaicataldo Yes, it's https://github.com/eslint/eslint/pull/8055

kaicataldo commented:
@mysticatea Aha! I thought I was going crazy, haha. Thanks!

kaicataldo commented:
Tests should pass and we should be ready to go after https://github.com/eslint/eslint/pull/8055 lands and I can rebase off of master.

kaicataldo commented:
This is now up to date and ready for review

mikesherov commented on lib/rules/lines-around-directive.js:
```diff
@@ -131,13 +131,14 @@ module.exports = {
-                firstDirective.leadingComments && firstDirective.leadingComments.length ||
+                leadingComments.length ||
 
                 // Shebangs are not added to leading comments but are accounted for by the following.
```
SEems like this comment is lying now?

kaicataldo replied:
Great catch - funny enough, I was the original author of that comment

kaicataldo commented on lib/eslint.js:
```diff
@@ -626,7 +625,7 @@ module.exports = (function() {
                 tokens: true,
                 comment: true,
-                attachComment: true,
+                attachComment: false,
```
Nope! We can remove this.

kaicataldo commented:
> Seeing the number of rules that skip shebang comments leads me to believe that the default should be to exclude, but that's just my two cents.

Yeah, as I was writing all those filters I had the thought that perhaps it would be nice to specify what types of comments `sourceCode.getAllComments()` returns (or, as you mentioned, excluding a shebang if it's present). Maybe we could add an opt-in option for when the shebang comment is desired (since I think that's the rare case)? Something like `sourceCode.getAllComments({ includeShebang: true })`?

>  I wonder if we can preserve visitor pattern for LineComment and BlockComment, I was unsure why those were removed.

The reason we removed visiting comments was that it was brought it up that changing the existing behavior could lead to more confusion than having to think about comments totally from a token iteration perspective ([comment here](https://github.com/eslint/eslint/pull/7516#issuecomment-272493694)).

> Seems like we can iterate over comments right before program:exit and preserve existing functionality.

I do think this is a viable alternative, but it would change the current behavior (since the comment events currently fire before and after the node to which they are attached). Though it would work, I'm not sure what firing all the events at once (unattached) buys a rule author over just calling `sourceCode.getAllComments()` and always thinking of comments from the perspective of token iteration rather than AST traversal.

mikesherov commented:
> @mikesherov Thanks for the review!

Seeing the number of rules that skip shebang comments leads me to believe that the default should be to exclude, but that's just my two cents.
Yeah, as I was writing all those filters I had the thought that perhaps it would be nice to specify what types of comments sourceCode.getAllComments() returns (or, as you mentioned, excluding a shebang if it's present). Maybe we could add an opt-in option for when the shebang comment is desired (since I think that's the rarer case)? Something like sourceCode.getAllComments({ includeShebang: true })?

I wonder if we can preserve visitor pattern for LineComment and BlockComment, I was unsure why those were removed.
The reason we removed visiting comments was that it was brought it up that changing the existing behavior could lead to more confusion than having to think about comments totally from a token iteration perspective (comment here).

Seems like we can iterate over comments right before program:exit and preserve existing functionality.
I do think this is a viable alternative, but it would change the current behavior (since the comment events currently fire before and after the node to which they are attached). Though it would work, I'm not sure what firing all the events at once (unattached) buys a rule author over just calling sourceCode.getAllComments() and always thinking of comments from the perspective of token iteration rather than AST traversal. Curious what you think the benefits are!

Sounds good. Don't want to reopen that can of worms!

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -906,11 +933,44 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

kaicataldo replied:
Good call!

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -906,11 +933,44 @@ describe("SourceCode", () => {
+                "if (/* Leading comment for Identifier */ a) {}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "var baz;"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "var b;"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "// Trailing comment for VariableDeclaration"
             ].join("\n");
 
             eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+        it("should include shebang comment when program only contains shebang", () => {
+            const code = "#!/usr/bin/env node";
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "};"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "/*another comment*/"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "})"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "]"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "}"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "// Trailing comment for VariableDeclaration"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

platinumazure commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,512 @@ describe("SourceCode", () => {
+                "    a;"
+            ].join("\n");
+
+            eslint.reset();
```
Can this `eslint.reset()` go in a `beforeEach`?

kaicataldo commented:
@btmills Really, really good call - there's definitely a bug here. Obvious now, but the logic of checking for "empty" nodes with comments isn't smart enough, which is why the leading comments aren't attached (it only attaches the trailing "inner" comments). Will update this this weekend!

not-an-aardvark commented:
Is the goal to make that all comments are attached somewhere? It might be possible to use a fuzzer to test that.

kaicataldo commented:
@not-an-aardvark No, after iterating on this for a while [we decided that it would be okay to lose some comments](https://github.com/eslint/eslint/pull/7516#issuecomment-259522555). 

I honestly think it makes more sense to think of comments in the context of token lists than in the context of AST nodes, and I think I'm actually leaning towards generally recommending using token iterator methods over sourceCode.getComments(). @nzakas mentioned [possibly deprecating it](https://github.com/eslint/eslint/pull/7516#issuecomment-272639595), though I think we can just recommend using the new token iterator methods instead of a `source.getCommentBefore|After`.

Part of my goal in all this is to unify and codify how we deal with comments so that it's simple and predictable. I'm not against updating things afterwards as long as we ensure it doesn't get overly complex (or there are too many different ways to do the same thing or slightly different things).

All that to say that I'm okay with not account for all comments with `sourceCode.getComments()`, because we currently have the leading/trailing paradigm. If we recommend `sourceCode.getAllComments()` as a way to get all comments and token iterator methods to get comments contextually/individually, I think we have a simple platform to build on in the future, and updating `sourceCode.getComments(node)` to simply return all comments inside the range of a node could be useful in a future major version.

not-an-aardvark commented on lib/eslint.js:
```diff
@@ -801,10 +798,7 @@ module.exports = (function() {
-                    shebang = captured;
-                    return `//${captured}`;
-                }),
+                stripUnicodeBOM(text).replace(/^#!([^\r\n]+)/, (match, captured) => `//${captured}`),
```
This will fail to parse a shebang that contains `\u2028` or `\u2029`. For example, Node is able to execute the following file without any problems, but eslint throws a parsing error. Note that there is a `\u2028` character between `bin` and `/env`.

```js
#!/usr/bin /env node
console.log("foo")
```

Actually, it seems like this was a problem before this PR, too. If there's no easy solution to this, we can probably just leave it.

kaicataldo replied:
Can you explain why they would be different, given that the regexp's are the same? What am I missing?

not-an-aardvark commented on lib/util/source-code.js:
```diff
@@ -160,6 +159,14 @@ function SourceCode(text, ast) {
     this.lines.push(this.text.slice(this.lineStartIndices[this.lineStartIndices.length - 1]));
 
+    // Check the source text for the presence of a shebang since it is parsed as a standard line comment.
+    const shebangTest = this.text.match(/^#!([^\r\n]+)/);
```
This is a correct match for shebangs, but unlike the logic in `lib/eslint.js`, the shebang match here will include `\u2028` and `\u2029` characters. If the logic in `lib/eslint.js` stays the way it is, this could result in some weird edge-cases where the shebang comment is also part of the AST if there is a `\u2028` or `\u2029` character in the shebang.

not-an-aardvark commented on docs/developer-guide/working-with-rules.md:
```diff
@@ -362,19 +362,27 @@ var nodeSourceWithFollowing = sourceCode.getText(node, 0, 2);
+### Accessing Shebangs
 
-**Note:** One of the libraries adds AST node properties for comments - do not use these properties. Always use `sourceCode.getComments()` as this is the only guaranteed API for accessing comments (we will likely change how comments are handled later).
+Shebangs are represented by tokens of type `"Shebang"`. They are treated as comments and can be accessed by the methods outlined above.
```
The `capitalized-comments` rule crashes at the moment when it encounters shebangs. It should probably ignore them instead.

kaicataldo replied:
This has been fixed!

kaicataldo commented on lib/rules/lines-around-comment.js:
```diff
@@ -207,12 +202,7 @@ module.exports = {
-            if (ancestors.length) {
-                parent = ancestors.pop();
-            }
+            const parent = sourceCode.getNodeByRangeIndex(node.range[0]);
```
I'm not sure about extracting this, since it really only works for comments - if you had an actual node's range it would return itself!

kaicataldo replied:
Alternatively, would a comment suffice?

btmills replied:
Yeah, that's a good point. And looks to be the only place we're using it, so a comment would work!

kaicataldo replied:
I ended up extracting it into its own method as well as renamed the incorrect `node` parameters to `token`. Hopefully this makes it all clearer, but Let me know if it doesn't :)

btmills replied:
I like it! That makes things plenty clear :+1:

kaicataldo commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,488 @@ describe("SourceCode", () => {
+                "    /* Leading comment for ExpressionStatement */",
+                "    foo();",
+                "    /* Trailing comment for ExpressionStatement */",
+                "}"
```
4 should be a leading comment for the BlockStatement - pushing a fix shortly!

kaicataldo commented on tests/lib/util/source-code.js:
```diff
@@ -928,19 +988,488 @@ describe("SourceCode", () => {
+                "switch (foo) {",
+                "    //comment",
+                "    /*another comment*/",
+                "}"
```
Okay, so looking into this, this is a case where thinking of comments in terms of nodes starts falling apart. For both `ClassDeclaration`s and `FunctionDeclaration`s, we're able to attach nodes to the `ClassBody` and `BlockStatement`, respectively. But in this case, we only have the two nodes - the `SwitchStatement` and the `Identifier`. Given the behavior in the other cases, I'm inclined to leave this behavior as is. What do you think?

kaicataldo replied:
Also wanted to mention that I'm proposing to actually deprecate `sourceCode.getComments()`: https://github.com/eslint/eslint/issues/8408

Ideally, my hope is that we can get `getComments()` in a place where it's good enough to not break existing rules currently relying on it. I don't think it's really feasible to make it have the exact same behavior as Espree's comment attachment strategy, and switching to thinking of comments in the context of token lists rather than AST nodes will make dealing with comments much more predictable!

btmills replied:
This all makes sense to me, and the new tests verify the behavior, let's go with it :+1:

kaicataldo commented:
@not-an-aardvark @btmills Thanks again for the thorough reviews! I have addressed all the comments (either with code changes or comments of my own). Please let me know what you think! Changes were made in the last two commits, so hopefully it's not too hard to re-review.

kaicataldo commented:
Also, rebased and ran eslint-canary against this branch - not seeing any new unexpected errors!

kaicataldo commented on tests/lib/rules/capitalized-comments.js:
```diff
@@ -80,6 +80,11 @@ ruleTester.run("capitalized-comments", rule, {
+        // Ignores shebangs
+        "#!/usr/bin/env node",
+        { code: "#!/usr/bin/env node", options: ["always"] },
+        { code: "#!/usr/bin/env node", options: ["never"] },
```
Good call - done!

not-an-aardvark commented:
Is the CLA bot down? It's still waiting for the status to be reported (and it hasn't left a comment)

vitorbal commented:
@not-an-aardvark that happened to me a couple of days ago. I had to force push to trigger the bot again.

JamesHenry commented:
Yayyyyy! What an epic PR 😄

Thanks so much for your work on this @kaicataldo! And to all the reviewers for their invaluable help.

mikesherov commented:
Congrats!
